### PR TITLE
Timestamp subtraction and interval operations for `ScalarValue`

### PR DESCRIPTION
**Pull request by berkaysynnada**
# Which issue does this PR close?

Closes #5411 and #5412.

# Rationale for this change

This PR is composed of two parts but served as one PR since the parts are closely related (they share constants, change similar places in the file). 340 lines are added for implementation, and 670 are for tests.

The first support is to timestamp subtraction within `ScalarValue`. In postgre, the behaviour is like: 
```
CREATE TABLE test (
  id INT,
  ts1 TIMESTAMP,
  ts2 TIMESTAMP
  
);
```
```
INSERT INTO test VALUES (1, '2016-03-13 00:00:00', '2017-04-13 12:11:03');
INSERT INTO test VALUES (2, '2016-03-04 06:00:00', '2016-03-01 20:33:48');
INSERT INTO test VALUES (3, '2017-03-04 23:59:59', '2016-03-01 12:00:00');
```
```
SELECT
ts2 - ts1 AS diff_ts
FROM test;
```
image

We can result in the same way with postgre. The difference is shown in days as the largest time unit. However, since we don't have such detailed fields for an interval, `TimestampSecond` or `TimestampMillisecond` timestamp subtractions give result in `IntervalDayTime` variant, and `TimestampMicrosecond` or `TimestampNanosecond` timestamp subtractions give result in `IntervalMonthDayNano` variant without using the month field. However, I need to underline that we can apply this operation only on scalar values. Supporting columnar operations is in the scope of arrow-rs, and it needs to be implemented in a similar way to have consistency.

The second support is comparing, adding and subtracting two `ScalarValue` interval types, `IntervalYearMonth`, `IntervalDayTime`, `IntervalMonthDayNano`. With this PR, we can apply these operations between both the same and different variants. arrow-rs support is required also here to support columnar values.
# What changes are included in this PR?

`impl_op` match arms are extended to cover timestamp and interval types. It should be noted that we need the same types of timestamps to apply subtraction. `impl PartialOrd for ScalarValue `and `impl PartialEq for ScalarValue` are extended to handle interval types. However, to be able to compare months and days, we need to assume a month equals to 30 days (postgre behaves in the same way).
# Are these changes tested?

Yes, there are some tests covering edge cases with timezone options, and randomized tests that first add/subtract an interval type to/from a timestamp, then take the difference between the resulting timestamp and the initial timestamp to assert the equality with the given interval. For the interval operations, some tests are written to cover all variations of operations for different types.

timestamp_next | timestamp_prev | Return as days
-- | -- | --
2017-04-16 00:00:00 . 000_000_100 | 2016-03-13 00:00:00 . 000_000_025 | months: 0, days: 399, nanos: 75
2016-03-13 00:00:00 . 000_000_025 | 2017-04-16 00:00:00 . 000_000_100 | months: 0, days: -399, nanos: -75
2016-12-16 01:00:00 . 100 | 2016-12-15 00:00:00 . 000 | days: 1, millis: 3600100
2016-12-15 00:00:00 . 000 | 2016-12-16 01:00:00 . 100 | days: -1, millis: -3600100

# Are there any user-facing changes?

-

# To run end to end queries including timestamp subtraction and interval operations:

1) `try_new_impl` function in arrow-rs gives an error while validating the schema and columns as such:

> `column types must match schema types, expected Timestamp(Second, None) but found Interval(DayTime) at column index 0`
 
That expectation of the output column needs to be changed to interval type.

.
2) In Datafusion, there is a function `coerce_types` returning the output type of applying `op` to an argument of `lhs_type` and `rhs_type`. These codes with the extensions in planner.rs and binary.rs also need to be reshaped. `evaluate` function in datetime.rs has to handle columnar values at each side.

### CODE REVIEW

File: datafusion/common/src/scalar.rs [hunk 1]
```diff
@@ -547,6 +572,272 @@ macro_rules! get_sign {
...
+// For example, lhs < rhs => Interval(-, -, -), lhs > rhs => Interval(+, +, +)
+fn ts_nanosec_sub_to_interval(
+    lhs_ts: &i64,
+    rhs_ts: &i64,
+    lhs_tz: &Option<String>,
+    rhs_tz: &Option<String>,
+) -> Result<ScalarValue, DataFusionError> {
+    // Conversion of integer and string-typed timestamps to NaiveDateTime objects
+    // Timezone offsets are added also if applicable.
+    let (naive_date_time2_unchecked, naive_date_time1_unchecked);
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:648]: Do not use variable without assignment.

File: datafusion/common/src/scalar.rs [hunk 2]
```diff
@@ -547,6 +572,272 @@ macro_rules! get_sign {
...
+    if let (Some(l), Some(r)) = (lhs_tz, rhs_tz) {
+        (naive_date_time2_unchecked, naive_date_time1_unchecked) =
+            integer_w_timezone_to_naive_datetime(lhs_ts, rhs_ts, l, r)?;
+    } else {
+        (naive_date_time2_unchecked, naive_date_time1_unchecked) =
+            integer_to_naive_datetime(lhs_ts, rhs_ts)?;
+    }
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:655]: Assign return value of the `if` `else` branches to variables `naive_date_time2_unchecked` and `naive_date_time1_unchecked`

File: datafusion/common/src/scalar.rs [hunk 3]
```diff
@@ -547,6 +572,272 @@ macro_rules! get_sign {
...
+}
+#[inline]
+fn find_interval_sign(
+    ndt2: NaiveDateTime,
+    ndt1: NaiveDateTime,
+) -> (NaiveDateTime, NaiveDateTime, i32) {
+    let sign;
+    if ndt2.timestamp_nanos() < ndt1.timestamp_nanos() {
+        sign = -1;
+        (ndt1, ndt2, sign)
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:770]: remove sign variable return `(ndt1, ndt2, -1)`

File: datafusion/common/src/scalar.rs [hunk 4]
```diff
@@ -547,6 +572,272 @@ macro_rules! get_sign {
...
+    } else {
+        sign = 1;
+        (ndt2, ndt1, sign)
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:773]: similar to the above

File: datafusion/common/src/scalar.rs [hunk 5]
```diff
@@ -547,6 +572,272 @@ macro_rules! get_sign {
...
+            if value == 0 {
+                return Ok(ScalarValue::IntervalYearMonth(Some(sign * months)));
+            }
+        }
+        None => {
+            return Err(DataFusionError::NotImplemented(String::from(
+                "months_residual nanosec overflow",
+            )))
+        }
+    }
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:677]: Instead of this use below
```rust
let value = months_residual.num_nanoseconds().ok_or_else(err)?;
if value == 0 {
    return Ok(ScalarValue::IntervalYearMonth(Some(sign * months)));
}
```

Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:677]: Instead of this use below
```rust
let err = || {
    DataFusionError::NotImplemented(String::from("months_residual nanosec overflow"))
};
// Check whether we can return an IntervalYearMonth variant without losing information
let value = months_residual.num_nanoseconds().ok_or_else(err)?;
if value == 0 {
    return Ok(ScalarValue::IntervalYearMonth(Some(sign * months)));
}
```

File: datafusion/common/src/scalar.rs [hunk 6]
```diff
@@ -547,6 +572,272 @@ macro_rules! get_sign {
...
+    // add it to months_residual to make it positive.
+    // To ensure the difference is positive all the time, we take the days
+    // of previous datetime's month.
+    if months_residual.num_nanoseconds() < Some(0) {
+        (months, months_residual) =
+            normalize_duration(&months, &months_residual, naive_date_time1)?;
+    }
+
+    // Check whether we can return an IntervalDayTime variant without losing information
+    let months_residual_in_ns = months_residual.num_nanoseconds().unwrap();
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:689]: remove this `.unwrap` like above using `.ok_or_else(err)`

File: datafusion/common/src/scalar.rs [hunk 7]
```diff
@@ -547,6 +572,272 @@ macro_rules! get_sign {
...
+    if months_residual_in_ns % 1_000_000 == 0 {
+        let delta_secs = naive_date_time2
+            .signed_duration_since(naive_date_time1)
+            .num_milliseconds();
+
+        return Ok(ScalarValue::IntervalDayTime(Some(
+            IntervalDayTimeType::make_value(
+                sign * (delta_secs / 86_400_000) as i32,
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:697]: Add a comment explaining 86_400_000 like (1 a day in miliseconds)

File: datafusion/common/src/scalar.rs [hunk 8]
```diff
@@ -547,6 +572,272 @@ macro_rules! get_sign {
...
+                sign * (delta_secs % 86_400_000) as i32,
+            ),
+        )));
+    }
+
+    Ok(ScalarValue::IntervalMonthDayNano(Some(
+        IntervalMonthDayNanoType::make_value(
+            sign * months,
+            sign * (months_residual_in_ns / 86_400_000_000_000) as i32,
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:706]: Similarly add a comment for this value

File: datafusion/common/src/scalar.rs [hunk 9]
```diff
@@ -547,6 +572,272 @@ macro_rules! get_sign {
...
+        (Some(l), Some(r)) => Ok((
+            DateTime::<FixedOffset>::from_utc(naive_lhs, l).naive_local(),
+            DateTime::<FixedOffset>::from_utc(naive_rhs, r).naive_local(),
+        )),
+        (_, _) => Ok((naive_lhs, naive_rhs)),
+    }
+}
+// This function parses as the format of "+HH:MM", for example, "+05:30"
+#[inline]
+fn parse_tz_to_offset(tz: &String) -> Option<FixedOffset> {
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:751]: Make this function return `Result` and remove `.unwrap`s in the function.

File: datafusion/common/src/scalar.rs [hunk 10]
```diff
@@ -547,6 +572,272 @@ macro_rules! get_sign {
...
+    let months_new = months - 1;
+    Ok((months_new, months_residual_new))
+}
+#[inline]
+// It gives the day count of the corresponding month at that year.
+fn days_in_month(year: i32, month: u32) -> Result<u32, DataFusionError> {
+    if let Some(first_day) = NaiveDate::from_ymd_opt(year, month, 1) {
+        let last_day = first_day
+            .with_month(month + 1)
+            .unwrap_or_else(|| NaiveDate::from_ymd_opt(year + 1, 1, 1).unwrap())
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:830]: Similarly remove this `.unwrap`

File: datafusion/common/src/scalar.rs [hunk 11]
```diff
@@ -547,6 +572,352 @@ macro_rules! get_sign {
...
+    match (lhs_ts.checked_mul(1_000), rhs_ts.checked_mul(1_000)) {
+        (Some(lhs_ns), Some(rhs_ns)) => {
+            ts_nanosec_sub_to_interval(&lhs_ns, &rhs_ns, lhs_tz, rhs_tz)
+        }
+        (None, _) => Err(DataFusionError::NotImplemented(format!(
+            "overflow while conversion of {lhs_ts:?}"
+        ))),
+        (_, None) => Err(DataFusionError::NotImplemented(format!(
+            "overflow while conversion of {rhs_ts:?}"
+        ))),
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:592]: You can replace code with below
```rust
let err_msg = "Overflow while conversion to microseconds";
let err = || DataFusionError::Execution(err_msg.to_string());
let lhs_ns = lhs_ts.checked_mul(1_000).ok_or_else(err)?;
let rhs_ns = rhs_ts.checked_mul(1_000).ok_or_else(err)?;
ts_nanosec_sub_to_interval(&lhs_ns, &rhs_ns, lhs_tz, rhs_tz)
```

File: datafusion/common/src/scalar.rs [hunk 12]
```diff
@@ -547,6 +572,352 @@ macro_rules! get_sign {
...
+    }
+}
+#[inline]
+fn ts_millisec_sub_to_interval(
+    lhs_ts: &i64,
+    rhs_ts: &i64,
+    lhs_tz: &Option<String>,
+    rhs_tz: &Option<String>,
+) -> Result<ScalarValue, DataFusionError> {
+    match (lhs_ts.checked_mul(1_000_000), rhs_ts.checked_mul(1_000_000)) {
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:602]: Do similar conversion with above

File: datafusion/common/src/scalar.rs [hunk 13]
```diff
@@ -547,6 +572,352 @@ macro_rules! get_sign {
...
+    }
+}
+#[inline]
+fn ts_sec_sub_to_interval(
+    lhs_ts: &i64,
+    rhs_ts: &i64,
+    lhs_tz: &Option<String>,
+    rhs_tz: &Option<String>,
+) -> Result<ScalarValue, DataFusionError> {
+    match (
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:621]: Do similar conversion

File: datafusion/common/src/scalar.rs [hunk 14]
```diff
@@ -547,6 +572,352 @@ macro_rules! get_sign {
...
+    // Check whether we will find a negative interval or not
+    let (naive_date_time2, naive_date_time1, sign) =
+        find_interval_sign(naive_date_time2_unchecked, naive_date_time1_unchecked);
+
+    // Subtraction of datetimes. Details are inside the function.
+    let (mut months, mut months_residual) =
+        datetime_month_sub_with_rem(naive_date_time2, naive_date_time1)?;
+
+    let err = || {
+        DataFusionError::NotImplemented(String::from("months_residual nanosec overflow"))
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:680]: The appropriate error type is not `NotImplemented` replace it with `Execution`. All error types are like this. Please check them all.

File: datafusion/common/src/scalar.rs [hunk 15]
```diff
@@ -4430,4 +4802,417 @@ mod tests {
...
+                        .and_hms_opt(0, 0, 0)
+                        .unwrap()
+                        .timestamp(),
+                ),
+                None,
+            ),
+        ]
+    }
+
+    fn new_timestamps_prev() -> Vec<ScalarValue> {
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:4957]: make name of this timestamps_prev

File: datafusion/common/src/scalar.rs [hunk 16]
```diff
@@ -4430,4 +4802,417 @@ mod tests {
...
+                        .and_hms_opt(0, 0, 0)
+                        .unwrap()
+                        .timestamp(),
+                ),
+                None,
+            ),
+        ]
+    }
+
+    fn new_expected_results_months(sign: i32) -> Vec<ScalarValue> {
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:5052]: make name of this `expected_results_months`

File: datafusion/common/src/scalar.rs [hunk 17]
```diff
@@ -4430,4 +4802,417 @@ mod tests {
...
+                sign * 425,
+                sign * 86370000,
+            ))),
+            ScalarValue::IntervalYearMonth(Some(IntervalYearMonthType::make_value(
+                sign * 43,
+                sign,
+            ))),
+        ]
+    }
+    fn new_expected_results_days(sign: i32) -> Vec<ScalarValue> {
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:5087]: make name of this `expected_results_days`

File: datafusion/common/src/scalar.rs [hunk 18]
```diff
@@ -4430,4 +4802,417 @@ mod tests {
             assert!(distance.is_none());
         }
     }
+
+    #[test]
+    fn timestamp_op_tests() {
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:4807]: You can move the randomized tests under this to a new test

File: datafusion/common/src/scalar.rs [hunk 19]
```diff
@@ -4430,4 +4802,417 @@ mod tests {
...
+                    IntervalMonthDayNanoType::make_value(month, day, nanosec),
+                )));
+            } else if i % 3 == 1 && i % 4 != 0 {
+                let day = rng.gen_range(0..=5000);
+                let millisec = rng.gen_range(0..86_400_000);
+                intervals.push(ScalarValue::IntervalDayTime(Some(
+                    IntervalDayTimeType::make_value(day, millisec),
+                )))
+            } else {
+                let year = rng.gen_range(0..=20);
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:5209]: In range you use 0..=20 why there is = in the range. You can use 0..20 I think. Is there any reason for it?

File: datafusion/common/src/scalar.rs [hunk 20]
```diff
@@ -547,6 +603,330 @@ macro_rules! get_sign {
...
+    ts_nanosec_sub_to_interval(&lhs_ns, &rhs_ns, lhs_tz, rhs_tz)
+}
+#[inline]
+fn ts_millisec_sub_to_interval(
+    lhs_ts: &i64,
+    rhs_ts: &i64,
+    lhs_tz: &Option<String>,
+    rhs_tz: &Option<String>,
+) -> Result<ScalarValue, DataFusionError> {
+    let err_msg = "Overflow while conversion to microseconds";
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:627]: Change error message to milliseconds

File: datafusion/common/src/scalar.rs [hunk 21]
```diff
@@ -547,6 +603,330 @@ macro_rules! get_sign {
...
+    ts_nanosec_sub_to_interval(&lhs_ns, &rhs_ns, lhs_tz, rhs_tz)
+}
+#[inline]
+fn ts_sec_sub_to_interval(
+    lhs_ts: &i64,
+    rhs_ts: &i64,
+    lhs_tz: &Option<String>,
+    rhs_tz: &Option<String>,
+) -> Result<ScalarValue, DataFusionError> {
+    let err_msg = "Overflow while conversion to microseconds";
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:640]: Change error message to seconds

File: datafusion/common/src/scalar.rs [hunk 22]
```diff
@@ -4486,4 +4744,423 @@ mod tests {
...
+    fn timestamp_op_random_tests() {
+        // timestamp1 + (or -) interval = timestamp2
+        // timestamp2 - timestamp1 (or timestamp1 - timestamp2) = interval ?
+        let sample_size = 10000000;
+        let timestamps1 = get_random_timestamps(sample_size);
+        let intervals = get_random_intervals(sample_size);
+        // ts(sec) + interval(ns) = ts(sec); however,
+        // ts(sec) - ts(sec) cannot be = interval(ns). Therefore,
+        // timestamps are more precise than intervals in tests.
+        let mut timestamp2: ScalarValue;
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:4787]: do not use without assigning variable

**Comment by mustafasrepo** [comment]
You need to resolve linter errors. Otherwise, it is good for me.

File: datafusion/common/src/scalar.rs [hunk 23]
```diff
@@ -4486,4 +4713,403 @@ mod tests {
...
+
+        for (idx, exp) in test_data.iter().enumerate() {
+            assert_eq!(exp.2, test_data[idx].0.sub(&test_data[idx].1).unwrap())
+        }
+
+        // negative interval, edge cases
+        let test_data = get_test_data(-1);
+        for (idx, exp) in test_data.iter().enumerate() {
+            assert_eq!(exp.2, test_data[idx].1.sub(&test_data[idx].0).unwrap());
+        }
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:4730]: replace body with code below
```rust
// positive interval, edge cases
let test_data = get_test_data(1);

for (lhs, rhs, expected) in test_data.iter() {
    assert_eq!(expected, &lhs.sub(rhs).unwrap())
}

// negative interval, edge cases
let test_data = get_test_data(-1);
for (rhs, lhs, expected) in test_data.iter() {
    assert_eq!(expected, &lhs.sub(rhs).unwrap());
}
```

File: datafusion/common/src/scalar.rs [hunk 24]
```diff
@@ -4486,4 +4713,403 @@ mod tests {
...
+                            .timestamp(),
+                    ),
+                    None,
+                ),
+                ScalarValue::IntervalDayTime(Some(IntervalDayTimeType::make_value(
+                    sign * 425,
+                    sign * 86370000,
+                ))),
+            ),
+            // 9. test case
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:4985]: replace with 9th. Do similar for all of the above comments

**Review by ozankabak — APPROVED** [review]
This LGTM. I checked in two commits: One that refactors and simplifies your code, and one that merges this PR with the other PR. Please study the changes in these commits and we will discuss any learnings tomorrow in our meeting. I closed the other PR as we will be moving forward with a single joint PR.

The rationale of going forward with a single joint PR is that my refactoring created merge conflicts with the other one. The PRs are/were closely related (they share constants, change similar places in the file etc.) so it is natural to send them in one batch. If upstream reviewers suggest breaking it into multiple PRs, we can break it down then.

The remaining work before sending upstream:
1. For directed test cases, change the comment lines to briefly explain which corner case you are testing (instead of just saying "Test Case 1, 2, ...").
2. If I am not misremembering our discussions, I think this does not close [#196](https://github.com/apache/arrow-datafusion/issues/194) as we also need the `arrow-rs` changes for the queries to work.
3. You need to write a new PR body that explains all the changes in the joint PR. Also, some of the things written in the current PR body looks old, it seems to require some updates to accurately reflect what is/will be in this PR. For example, this PR has only a part of the work required to run queries end to end and this has to be clear. We should also briefly discuss what the remaining steps are for this goal.

You can send it upstream after making these changes.
